Src/Base/AMReX_GpuFuse.cpp

```diff
@@ -0,0 +1,241 @@
+#include <AMReX_Gpu.H>
+#include <AMReX_ParmParse.H>
+#include <AMReX_BLProfiler.H>
+#include <AMReX_OpenMP.H>
+
+namespace amrex {
+namespace Gpu {
+
+#ifdef AMREX_USE_CUDA
+
+namespace {
+    bool s_in_fuse_region = false;
+    // Fusing kernels with elements greater than this are not recommended based tests on v100
+    Long s_fuse_size_threshold = 257*257;
+    // If the number of kernels is less than this, fusing is not recommended based on tests on v100
+    int s_fuse_numkernels_threshold = 3;
+}
+
+std::unique_ptr<Fuser> Fuser::m_instance = nullptr;
+
+Fuser::Fuser ()
+{
+    AMREX_ASSERT(!OpenMP::in_parallel());
+    m_lambda_buf = (char*)The_Pinned_Arena()->alloc(m_nbytes_lambda_buf);
+    m_helper_buf = (FuseHelper*)The_Pinned_Arena()->alloc(m_nhelpers_buf*sizeof(FuseHelper));
+    m_dtor_buf.reserve(1024);
+}
+
+Fuser::~Fuser ()
+{
+    if (m_nlambdas > 0) {
+        Launch();
+    }
+
+    The_Pinned_Arena()->free(m_lambda_buf);
+    The_Pinned_Arena()->free(m_helper_buf);
+}
+
+void Fuser::Launch ()
+{
+    BL_PROFILE("Fuser::Launch()");
+
+    int nlambdas = m_nlambdas;
+    if (nlambdas > 0) {
+        AMREX_ASSERT(!OpenMP::in_parallel());
+
+        int* nwarps = (int*)The_Pinned_Arena()->alloc((nlambdas+1)*sizeof(int));
+        int ntotwarps = 0;
+        for (int i = 0; i < nlambdas; ++i)
+        {
+            nwarps[i] = ntotwarps;
+            Box const& bx = m_helper_buf[i].m_bx;
+            int N;
+            if (bx.isEmpty()) {
+                N = m_helper_buf[i].m_N;
+            } else {
+                N = bx.numPts();
+            }
+            ntotwarps += (N + Gpu::Device::warp_size-1) / Gpu::Device::warp_size;
+        }
+        nwarps[nlambdas] = ntotwarps;
+
+        AMREX_ASSERT(ntotwarps < std::numeric_limits<int>::max()/Gpu::Device::warp_size);
+
+        // Pack nwarps, lambda helpers and lambda objects into a buffer
+        std::size_t sizeof_nwarps = sizeof(int) * (nlambdas+1);
+        std::size_t offset_helpers = Arena::align(sizeof_nwarps);
+        std::size_t sizeof_helpers = sizeof(FuseHelper)*nlambdas;
+        std::size_t offset_objects = Arena::align(offset_helpers+sizeof_helpers);
+        std::size_t sizeof_objects = m_nbytes_used_lambda_buf;
+        std::size_t total_buf_size = offset_objects + sizeof_objects;
+
+        char* h_buffer = (char*)The_Pinned_Arena()->alloc(total_buf_size);
+        char* d_buffer = (char*)The_Device_Arena()->alloc(total_buf_size);
+
+        std::memcpy(h_buffer, nwarps, sizeof_nwarps);
+        std::memcpy(h_buffer+offset_helpers, m_helper_buf, sizeof_helpers);
+        std::memcpy(h_buffer+offset_objects, m_lambda_buf, sizeof_objects);
```

Comment:
As @WeiqunZhang and I discussed offline, this seems to work even though the extended host / device lambda is not technically trivial copyable. @WeiqunZhang convinced me that this will probably have to keep working in the future, but maybe it's something we should keep an eye on.

Reply:
I talked to our compiler team about this and let them know that they should keep an eye on this too. I think this particular usage is probably OK.

Reply:
Thanks!